mitodl/ocw-studio
# Youtube metadata update

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #545

#### What's this PR do?
- Autopopulates the youtube thumbnail field for video resources of ocw sites if/when the Youtube ID field is manually populated.
- Checks to see if any video resources have unassigned youtube ids before previewing/publishing an OCW site.  Previews will proceed but a warning will be sent in the response.  Publishing will be aborted and an error will be returned in the response.
- Updates youtube metadata for all of an OCW website's video resources when a preview/publish request is made.  This is done first, so that if it fails (ie quota limit exceeded), then the rest of the preview/publish process is aborted.
- Front end feedback for warnings/errors is not included, it will be done as part of #386 

#### How should this be manually tested?
Testing successful youtube uploads and updates will require setting up a personal account for now, the one used for RC will always fail due to a daily quota limit of 0.  How to do so is described in the README, and you can follow the instructions for testing uploads in PR #484.

Once you have videos uploaded to Youtube, you can test the functionality in this PR as follows:
- Set `OCW_IMPORT_STARTER_SLUG=ocw-course` in your `.env` file (and make sure you have a starter with that slug, same config as on RC).
- Create a new video resource for a website created with the `ocw-course` starter, and fill in the `Youtube ID` field with the id of one of your uploaded youtube videos.  Also fill in the description and the "Youtube Speakers" and 'Youtube Tags' fields. Then save.
- Edit the above resource.  Check that the "Video Thumbnail URL" field is populated with `https://img.youtube.com/vi/<youtube_id>/0.jpg`.  This URL actually won't work yet on your personal account because all your videos are being uploaded as private due to your app not being verified.  This should hopefully not be an issue on RC or production.
- With your network tab open in your browser, click 'Preview' and 'Publish'.  Both should return 200 responses with no warnings about missing youtube ids (assuming you only have this 1 video resource for the website).
- Check the video you uploaded to youtube.  Your video resource title should now be the youtube title, the description should be a combination of the 'Description' and 'Youtube Speakers' fields, and the youtube tags should equal your video resource tags.  Your video will still be private due to your app being unverified.
- Create another video resource with a blank Youtube ID.  Preview and publish again.  The preview response should still be a 200, but with a warning about 1 of your videos not having a Youtube ID.  The publish response should have a 400 status with an error about that video.
- Try to repeat the above with a different website based on a different starter.  The preview and publish responses should always return a 200 status.
